### PR TITLE
Multiple improvements to PPO

### PR DESCRIPTION
**Pull request by liujch1998**
Improvements:
1. Breaking down rollout / forward / backward into separate JIT functions, which saves TPU memory.
1. Fixed the logging of time so that it is now accurate, and log the time spent on each part (rollout, forward, backward, total).
1. Support using bigger `batch_size` than `forward/backward_mini_batch_size`. (Fixed some sharding issue)
1. Gradient accumulation now works properly.
1. Support multi-rollout by specifying `rollouts_per_prompt`.
1. Support freezing policy at beginning of training by specifying `policy_freeze_epochs`. (For some reason, converging the value loss is needed when using bigger `batch_size`, otherwise the policy will degenerate)
1. Removed the trainstate wrapper for ref_policy and RM, which presumably removes the optimizer states allocated for them. This saved ~33% TPU memory.

I renamed some older TPU kickoff scripts with prefix `old_*.sh`. These are not updated with the latest hparams so don't use them unless you update them.

Example run: <https://wandb.ai/liujch1998/n-Tulu-PPO-Jax/runs/train_v3_v2.5.1_ppo3_cleaned-uf-data--ead95f8f135b4f00886582003ae552db>

Time logging:
Screenshot 2024-04-17 at 21 16 55

Verified that this yields roughly the same outcome with our previous setup (batch_size=64): <https://wandb.ai/liujch1998/n-Tulu-PPO-Jax/runs/train_v3.0.4_v3_replicating-v2_b64--c34b91ba4da44b97b7957054e17255b2>
Baseline for comparison: <https://wandb.ai/rlhf-llm-dev/easylm_ppo/runs/tulu_2_base_rm_ultrafeedback_60k--43857cd946dc4d4fadd17d17b7a49d60>

### CODE REVIEW

File: EasyLM/checkpoint.py [hunk 1]
```diff
@@ -66,8 +66,9 @@ def save_pickle(self, obj, filename):
             path = '/dev/null'
         mlxu.save_pickle(obj, path)
 
-    def save_all(self, train_state, gather_fns, metadata=None, dataset=None, milestone=False, is_value=False):
-        step = int(jax.device_get(train_state.step))
+    def save_all(self, train_state, gather_fns, metadata=None, dataset=None, milestone=False, step=None, is_value=False):
+        if step is None:
+            step = int(jax.device_get(train_state.step))
```
Comment (liujch1998) [inline EasyLM/checkpoint.py:71]: By default the checkpointer would count steps as the number of backward passes (i.e. `train_state.step`). This is quite confusing so I wrote an option to override it.

File: EasyLM/models/llama/llama_train_ppo.py [hunk 2]
```diff
@@ -36,9 +36,8 @@
 
 FLAGS, FLAGS_DEF = mlxu.define_flags_with_default(
     seed=42,
-    initialize_jax_distributed=False,
```
Comment (liujch1998) [inline EasyLM/models/llama/llama_train_ppo.py:39]: this flag is not used anywhere here

File: EasyLM/models/llama/llama_train_ppo.py [hunk 3]
```diff
@@ -80,8 +83,14 @@
...
     max_steps_per_epoch=0,
     generate_only=False,
+    no_backward=False,
 )
 
+pad_token_id = 0
+bos_token_id = 1
+eos_token_id = 2
+
+
```
Comment (liujch1998) [inline EasyLM/models/llama/llama_train_ppo.py:94]: Keeping those hard-coded here. There's additional complication with tokenizers and let's review this later.

Comment (hamishivi) [reply EasyLM/models/llama/llama_train_ppo.py:94]: With llama 3 (new bos/eos id), this might cause issues. happy to merge now but we should fix this in the near future.

**Review by hamishivi — APPROVED** [review]
Seems good!

File: EasyLM/models/llama/llama_train_ppo.py [hunk 4]
```diff
@@ -584,34 +585,30 @@ def save_checkpoint(policy_train_state, value_train_state, step, milestone=False
...
-            if reference_params is None:
-                reference_train_state = sharded_init_fn_policy(next_rng())
-            else:
-                if not FLAGS.use_tpu:
-                    reference_params = flax.core.frozen_dict.unfreeze(reference_params)
-                reference_train_state = sharded_create_trainstate_from_params_policy(reference_params)
-                del reference_params
+            assert reference_train_state is None
+        if reference_params is None:
+            reference_params = copy.deepcopy(policy_train_state.params)
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py:595]: does this work correctly on tpu?

Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py:595]: Yep. I'm using this for my experiments now, seems no problem

**Comment by hamishivi** [comment]
Thanks so much!
